### PR TITLE
[extractor/bandcamp] Add album_artist metadata (#5536)

### LINKED ISSUE #5536: [bandcamp.com] add album_artist metadata field for tracks
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a site-specific feature
- [X] I've verified that I'm running yt-dlp version **2022.11.11** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

USA and EU at least, pretty much global I believe.

### Example URLs

Any album with tracks of differing artists, for example:
https://damnably.bandcamp.com/album/damnably-famnably-14th-birthday-free-compilation
https://isolatedsw.bandcamp.com/album/isolated-free-compilation-vol-1
https://diskotopia.bandcamp.com/album/dsk-f-w-2016-2017-free-compilation

### Provide a description that is worded well enough to be understood

Bandcamp tracks currently do not include the `album_artist` metadata field which makes downloading compilations difficult specifically when the user desires the downloaded album to be grouped together in a directory under a single artist name, a common pattern for media servers like [Jellyfin](https://jellyfin.org/):
```
.
└── artist name
    └── album name
        ├── 1 foo.mp3
        ├── 2 bar.mp3
        └── 3 baz.mp3
```
For albums containing tracks where all the artists are identical, this can be done easily with `--format %(artist)s/%(album)s/%(track_number)s %(track)s.%(ext)s`. But for compilation albums (albums with tracks of differing artists), the album will be separated into many `artist name /album name` directories which breaks apart the album and is not good for Jellyfin and other media servers because they will see it as multiple albums with the same name.

What I'm proposing is that the bandcamp extractor include an `album_artist` field so a user can use the above directory layout scheme even with compilation albums by using `--format %(album_artist)s/%(album)s/%(track_number)s %(track)s.%(ext)s`. Bandcamp already associates a single artist which owns each album, but that artist information isn't provided in each track's JSON metadata so it needs to be scraped from the HTML of the track webpage:

2022-11-14-10:10:15-screenshot

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', '--output', 'ignore/%(artist)s/%(track_number)s %(track)s.%(ext)s', 'https://damnably.bandcamp.com/album/damnably-famnably-14th-birthday-free-compilation']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.11.11 [8b644025b] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: d7b460d0e
[debug] Python 3.10.7 (CPython x86_64 64bit) - Linux-5.19.16-200.fc36.x86_64-x86_64-with-glibc2.35 (OpenSSL 3.0.5 5 Jul 2022, glibc 2.35)
[debug] exe versions: ffmpeg 5.0.1 (setts), ffprobe 5.0.1
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.1
[debug] Proxy map: {}
[debug] Loaded 1725 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.11.11, Current version: 2022.11.11
yt-dlp is up to date (2022.11.11)
[debug] [Bandcamp:album] Extracting URL: https://damnably.bandcamp.com/album/damnably-famnably-14th-birthday-free-compilation
[Bandcamp:album] damnably-famnably-14th-birthday-free-compilation: Downloading webpage
[download] Downloading playlist: Damnably Famnably 14th Birthday FREE Compilation
[Bandcamp:album] Playlist Damnably Famnably 14th Birthday FREE Compilation: Downloading 22 items of 22
[download] Downloading item 1 of 22
[debug] [Bandcamp] Extracting URL: https://damnably.bandcamp.com/track/old-town
[Bandcamp] old-town: Downloading webpage
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 3426938226: Downloading 1 format(s): mp3-128
[debug] Invoking http downloader on "https://t4.bcbits.com/stream/ad55924dcf5d5c30270a0643d30887ee/mp3-128/3426938226?p=0&ts=1668536489&t=8dca28d1becd6b438721c91c639a6f3f8020b3c8&token=1668536489_76a7045038a78721586c55213e322ff8175eb24d"
[download] Destination: ignore/Say Sue Me/1 Say Sue Me - Old Town.mp3
[download] 100% of    3.56MiB in 00:00:01 at 3.53MiB/s
[download] Downloading item 2 of 22
[debug] [Bandcamp] Extracting URL: https://damnably.bandcamp.com/track/dont-light-my-fire
[Bandcamp] dont-light-my-fire: Downloading webpage
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 2553803020: Downloading 1 format(s): mp3-128
[debug] Invoking http downloader on "https://t4.bcbits.com/stream/0ac3265b9ff5685ed77df0c806a43b61/mp3-128/2553803020?p=0&ts=1668536491&t=8b2551f375287c9f9e8089a7bfb1532ce731fd73&token=1668536491_26025717c9fadcce2250a1e533101465e1c3e133"
[download] Destination: ignore/Otoboke Beaver/2 Otoboke Beaver - Don't Light My Fire.mp3
[download] 100% of    2.09MiB in 00:00:00 at 3.93MiB/s
[download] Downloading item 3 of 22
[debug] [Bandcamp] Extracting URL: https://damnably.bandcamp.com/track/national-police-shit-2
[Bandcamp] national-police-shit-2: Downloading webpage
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 1007927037: Downloading 1 format(s): mp3-128
[debug] Invoking http downloader on "https://t4.bcbits.com/stream/0a7e36da45aa922ca52cfe8cb12961f3/mp3-128/1007927037?p=0&ts=1668536493&t=98fc6dc7960528ffad423abe9762bb9e25cd024f&token=1668536493_634902abb1ca63aaf18e6b5e2258eb4f3b82cbb0"
[download] Destination: ignore/Drinking Boys and Girls Choir/3 Drinking Boys and Girls Choir - National Police Shit.mp3
[download] 100% of    1.42MiB in 00:00:01 at 957.48KiB/s
[download] Downloading item 4 of 22
[debug] [Bandcamp] Extracting URL: https://damnably.bandcamp.com/track/dream-grrrl-single-version-2
[Bandcamp] dream-grrrl-single-version-2: Downloading webpage
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 3278203047: Downloading 1 format(s): mp3-128
[debug] Invoking http downloader on "https://t4.bcbits.com/stream/43a05c30dcaaad31dfaf57e92a104a2f/mp3-128/3278203047?p=0&ts=1668536495&t=0b3ecc6fb85bc6a8db4708fdeda28265e6a262c6&token=1668536495_7ff6d10331e95f09bcebdde48c86ed377a3669d3"
[download] Destination: ignore/Grrrl Gang/4 Grrrl Gang - Dream Grrrl (Single Version).mp3
[download] 100% of    3.14MiB in 00:00:01 at 2.11MiB/s
[download] Downloading item 5 of 22
[debug] [Bandcamp] Extracting URL: https://damnably.bandcamp.com/track/teenage-wasteland
[Bandcamp] teenage-wasteland: Downloading webpage
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 2875477778: Downloading 1 format(s): mp3-128
[debug] Invoking http downloader on "https://t4.bcbits.com/stream/feff7121725401cac13cb5a8b752259d/mp3-128/2875477778?p=0&ts=1668536497&t=fcf2708340d75aaa2694b8c5d22a722783592f62&token=1668536497_2a2a8b07fe0e2ec81e6ff120b55ca385c830bbac"
[download] Destination: ignore/Wussy/5 Wussy - Teenage Wasteland.mp3
[download] 100% of    4.27MiB in 00:00:00 at 4.60MiB/s
```

### PR DESCRIPTION
**Pull request by stelcodes**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

This patch adds an `album_artist` metadata field in the `BandcampIE` extractor class by scraping the track's webpage for the name of the artist who owns the album. This makes it easier to download albums into an `artist/album/track.mp3` directory structure when downloading compilation albums (tracks of varying artists).

Fixes #5536 

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/bandcamp.py [hunk 1]
```diff
@@ -84,6 +84,11 @@ def _extract_data_attr(self, webpage, video_id, attr='tralbum', fatal=True):
             r'data-%s=(["\'])({.+?})\1' % attr, webpage,
             attr + ' data', group=2), video_id, fatal=fatal)
 
+    def _extract_album_artist(self, webpage, fatal=False):
+        return self._html_search_regex(
+            r'<h3 class="albumTitle">[\s\S]*by\s*<span>\s*<a href=".*">\s*(.*)\s*</a>',
```
Comment (pukkandan) [inline yt_dlp/extractor/bandcamp.py:89]:
```suggestion
            r'<h3 class="albumTitle">[^>]*by\s*<span>\s*<a href="[^>]+">\s*([^>]+?)\s*</a>',
```

File: yt_dlp/extractor/bandcamp.py [hunk 2]
```diff
@@ -84,6 +84,11 @@ def _extract_data_attr(self, webpage, video_id, attr='tralbum', fatal=True):
...
+            webpage, "album artist", group=1, fatal=fatal)
```
Comment (pukkandan) [inline yt_dlp/extractor/bandcamp.py:90]:
```suggestion
            webpage, 'album artist', fatal=False)
```

Comment (pukkandan) [inline yt_dlp/extractor/bandcamp.py:90]: This does not need to be a function

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/bandcamp.py (hunk 1 above)
Comment (stelcodes) [reply yt_dlp/extractor/bandcamp.py:89]: Ah nice I like that better

**Comment by pukkandan** [comment]
No. I will squash everything before merge. So commit messages don't really matter

File: yt_dlp/extractor/bandcamp.py [hunk 3]
```diff
@@ -121,6 +121,11 @@ def _real_extract(self, url):
         embed = self._extract_data_attr(webpage, title, 'embed', False)
         current = tralbum.get('current') or {}
         artist = embed.get('artist') or current.get('artist') or tralbum.get('artist')
+        with open('ignore/damnably-webpage-example.html', 'w') as f:
+            f.write(webpage)
```
Comment (pukkandan) [inline yt_dlp/extractor/bandcamp.py:125]: ?

**Comment by stelcodes** [comment]
@pukkandan Oh jeez I committed the wrong bit. That's embarrassing. Fixed!

**Comment by stelcodes** [comment]
Added myself to the contributors doc too if that's alright.

**Comment by stelcodes** [comment]
@pukkandan ping - just letting you know PR is ready for review again

**Comment by pukkandan** [comment]
Pls update tests

**Comment by stelcodes** [comment]
Updated to include tests in the `BandcampIE` class. Running `python test/test_download.py TestDownload.test_Bandcamp` locally gave me errors for pretty much every other test in the `_TESTS` list.  Confusing but I just commented them out and wrote a test that covers a track with a differing artist/album_artist that passes.

**Comment by pukkandan** [comment]
The test suite wasn't as strict till recently. So many of the tests are incomplete by current standards. It's expected to be fixed when we work on the extractor. Also, due to the nature of extractors, tests often break due to factors outside our control. In that case, we have to manually check and fix the test definitions.

**Pull request merged**